src/main/java/org/tugraz/sysds/runtime/lineage/RewriteCPlans.java

```diff
@@ -0,0 +1,229 @@
+import org.tugraz.sysds.runtime.matrix.data.MatrixBlock;
+import org.tugraz.sysds.runtime.meta.MetaData;
+import org.tugraz.sysds.utils.Explain.ExplainType;
+
+public class RewriteCPlans
+{
+	private static final String LR_VAR = "__lrwrt";
+	private static BasicProgramBlock _lrPB = null;
+	private static ExecutionContext _lrEC = null;
+	
+	public static boolean executeRewrites (Instruction curr, ExecutionContext ec)
+	{
+		boolean oneappend = false;
+		boolean twoappend = false;
+		MatrixBlock lastResult = null;
+		if (curr.getOpcode().equalsIgnoreCase("tsmm"))
+		{
+		// If the input to tsmm came from cbind, look for both the inputs in cache.
+			LineageItem[] items = ((ComputationCPInstruction) curr).getLineageItems(ec);
+			LineageItem item = items[0];
+
+			for (LineageItem source : item.getInputs())
+				if (source.getOpcode().equalsIgnoreCase("append")) {
+					for (LineageItem input : source.getInputs()) {
+						// create tsmm lineage on top of the input of last append
+						LineageItem tmp = new LineageItem("toProbe", curr.getOpcode(), new LineageItem[] {input});
+						if (LineageCache.probe(tmp)) {
+							oneappend = true;            // at least one entry to reuse
+							if (lastResult == null)
+								lastResult = LineageCache.get(curr, tmp);
+						}
+					}
+					if (oneappend)
+						break;   // no need to look for the next append
+
+					// if not found in cache, look for two consecutive cbinds
+					LineageItem input = source.getInputs()[0];
+					if (input.getOpcode().equalsIgnoreCase("append")) {
+						for (LineageItem L2appin : input.getInputs()) {
+							LineageItem tmp = new LineageItem("comb", "append", new LineageItem[] {L2appin, source.getInputs()[1]});
+							LineageItem toProbe = new LineageItem("toProbe", curr.getOpcode(), new LineageItem[] {tmp});
+							if (LineageCache.probe(toProbe)) {
+								twoappend = true;
+								if (lastResult == null)
+									lastResult = LineageCache.get(curr, toProbe);
+							}
+						}
+					}
+				}
+		}
+		else
+			return false;
+
+		if (!oneappend && !twoappend) 
+			return false;
+		
+		ExecutionContext lrwec = getExecutionContext();
+		ArrayList<Instruction> newInst = null;
+
+		try {
+		if (oneappend)
+			newInst = rewrite1(curr, ec, lrwec, lastResult);
+		if (twoappend)
+			newInst = rewrite2(curr, ec, lrwec, lastResult);
+		//execute instructions
+		BasicProgramBlock pb = getProgramBlock();
+		pb.setInstructions(newInst);
+		LineageCacheConfig.shutdownReuse();
+		pb.execute(lrwec);
+		LineageCacheConfig.restartReuse();
+		
+		ec.setVariable(((ComputationCPInstruction)curr).output.getName(), lrwec.getVariable(LR_VAR));
+		// add this to cache
+		LineageCache.put(curr, ec);
+		}
+		catch (Exception e) {
+			throw new DMLRuntimeException("Error evaluating instruction: " + curr.toString() , e);
+		}
+		return true;
+	}
+
+	static ArrayList<Instruction> rewrite1(Instruction curr, ExecutionContext ec, ExecutionContext lrwec, MatrixBlock lastResult) 
```

Comment:
please use better names than rewrite1 - you might want to take a look at our algebraic simplification rewrites.